package compact_descriptor;

message Description {
  repeated Component component = 1;
}

message Component {
  enum Type {
    Rack = 0;
    Module = 1;
    Knob = 2;
    Selector = 3;
    ValueInputPort = 4;
    ValueOutputPort = 5;
    GateInputPort = 6;
    GateOutputPort = 7;
  }

  required string name = 1;
  required Type type = 2;
  required uint32 id = 3;
  repeated Attribute attribute = 4;
  repeated Component sub_component = 5;
}

message Attribute {
  enum Type {
    Value = 0;
    Scale = 1;
    Choices = 2;
    WireId = 3;
    Direction = 4;
    Signal = 5;
  }

  required Type type = 1;
  optional uint32 ivalue = 2;
  repeated string svalue = 3;
}
